### ISSUE TITLE
QrCode - a little investigation

### ISSUE THREAD

**Issue by OlofSvahnVbg** [hajkmap/Hajk#1407]
Someone had the idea of adding a qr-code option for print in kommungis. The idea is to have an option for generating a qr-code here:

![image](https://github.com/hajkmap/Hajk/assets/111860819/36ef592c-e043-4fdb-92f7-795f3b86f382)

The option would be initially turned off, as to not overlap with logo, scalebar and north arrow. If turned on, the warning for overlapping icons would work the same, of course.

If chose correctly, however, the qr-code would appear like this:
![image](https://github.com/hajkmap/Hajk/assets/111860819/ee8f9410-03a8-4c57-b2ec-8a16f368b0f2)

The qr-code is implemented by using `QrCode from "qrcode`. 
Link to library: https://www.npmjs.com/package/qrcode

And a possible logic for this in `PrintModel` would be this:

`  generateQR = async () => {
    try {
      const url = AppModel.anchorModel.getAnchor();
      console.log(url);
      return await QRCode.toDataURL(url);
    } catch (err) {
      return "";
    }
  };`

`      //If-statement for checking qr-code option
      if (options.includeQrCode) {
        try {
          const qrCodeImageData = await this.generateQR();
          console.log(qrCodeImageData);

          let qrCodePlacement = this.getPlacement(
            options.qrCodePlacement,
            20,
            20,
            dim[0],
            dim[1]
          );

          pdf.addImage(
            qrCodeImageData,
            "PNG",
            qrCodePlacement.x,
            qrCodePlacement.y,
            20,
            20
          );
        } catch (error) {
          // The image loading may fail due to e.g. wrong URL, so let's catch the rejected Promise
          this.localObserver.publish("error-loading-qr-image");
        }
      }`

However, in testing this, I encountered many problems with generating and updating the url. The `AppModel.anchorModel.getAnchor()` is not enough for correctly generating a updated url with every layer and and zoom an so on. I guess you would have to copy much of the logic from `Anchor` to keep it updated. 

I'm not sure what the best solution is. Further investigation is needed.

In conclusion, I would say this is a cool feature that is possible to implement. But, the question is if it's worth the trouble. Perhaps not.

**Comment by jacobwod** [comment]
Great idea.

I would suggest we use the functionality added in #1252 for this. Start by enabling `enableAppStateInHash` in all your maps (it's an Admin setting). (Ideally, in future releases, this will be made the default functionality. I'm open for enabling it by default. It's stable, we've been using it for 6+ months in production.)

With that enabled, **you generate the QR code by taking the current value from `window.location.href`**. That's it. 🚀

**Comment by OlofSvahnVbg** [comment]
Awesome! Thanks. I'll take a look at it.

Den tis 3 okt. 2023 kl 11:04 skrev Jacob Wodzyński ***@***.***

**Comment by Hallbergs** [comment]
I think it would be worth the trouble, QR-codes are cool!

**Comment by jacobwod** [comment]
I read this again and saw this from your initial post:
> However, in testing this, I encountered many problems with generating and updating the url. The `AppModel.anchorModel.getAnchor()` is not enough for correctly generating a updated url with every layer and and zoom an so on. I guess you would have to copy much of the logic from `Anchor` to keep it updated.

Are you sure about this? I had a look at this and the [`getAnchor` method](https://github.com/hajkmap/Hajk/blob/f4222b10ef104801723c166972475c585beed837/new-client/src/models/AnchorModel.js#L162) does take all the map View's parameters into account, as well as layers (including group layers), plus current query string in search box, currently visible plugins and current CQL filters (which nobody seems to use anyway 😉). 

What else do you need? 😄

**Comment by OlofSvahn** [comment]
Yo yo! Im back, and here's a little update on qr code: 

Turns out the enableAppStateInHash worked perfectly. Me and Jesper talked about checking for this value to be true before enabling qr code client side. Otherwise it doesnt work with simply using the updated url.

**Comment by OlofSvahn** [comment]
Thanks! Good to be back :) Maybe we dont need to check for enableAppState then?

**Comment by jacobwod** [comment]
I think it's still a good idea to check, just as it's currently done elsewhere, see e.g. https://github.com/hajkmap/Hajk/blob/1087aaaa2df1a0697c9b84e7277513cd5c12ee4b/new-client/src/components/App.js#L576.
